### PR TITLE
Fix dependency loop between INITRAMFS_IMAGE and UKI

### PR DESCRIPTION
**Pull request by quic-vkraleti**
Ramdisk image by default pulls in a dependency on do_packagedata task of kernel.
Uki.bbclass is trying to create kernel-uki package, by packaging ramdisk image before packagedata task. 
This is the causing circular dependency of tasks. To address this, need to stop creating kernel-uki package and place efi image in deploydir instead.

### CODE REVIEW

**Comment by lumag** [comment]
We are back to picking up the image from deploy dir. I think it might be better to package the UKI separately from its own recipe. This way it will not be limited to just kernel+initramfs, but will simplify extending it with the splash screen, microcode, etc.

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+COMPATIBLE_HOST = '(x86_64.*|arm.*|aarch64.*)-(linux.*)'
+
+inherit uki
+
+do_configure[noexec] = "1"
+
+python do_compile() {
+    bb.build.exec_func('do_uki', d)
+}
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:14]: Just rename do_uki to do_compile

File: classes/uki.bbclass [hunk 2]
```diff
@@ -13,17 +13,18 @@ DEPENDS:append = " \
...
+    virtual/kernel:do_deploy \
+    ${INITRAMFS_IMAGE}:do_image_complete \
     "
-do_install[depends] += "${@ '${INITRAMFS_IMAGE}:do_image_complete' if d.getVar('INITRAMFS_IMAGE') else ''}"
-do_uki[dirs] = "${B}"
-do_install[postfuncs] += "do_uki"
 
-python do_uki() {
-    import glob
+fakeroot python do_uki() {
```
Comment (lumag) [inline classes/uki.bbclass:25]: Why do you need fakeroot here?

File: classes/uki.bbclass [hunk 3]
```diff
@@ -13,17 +13,18 @@ DEPENDS:append = " \
...
 
-inherit python3native
+inherit python3native image-artifact-names
 
-do_install[depends] += " \
+do_configure[depends] += " \
+    os-release:do_populate_sysroot \
     systemd-boot:do_deploy \
+    virtual/kernel:do_deploy \
+    ${INITRAMFS_IMAGE}:do_image_complete \
```
Comment (lumag) [inline classes/uki.bbclass:22]: This will break if `INITRAMFS_IMAGE` is not defined.

File: classes/uki.bbclass [hunk 4]
```diff
@@ -33,23 +34,34 @@ python do_uki() {
...
-        baseinitrd = os.path.join(d.getVar('INITRAMFS_DEPLOY_DIR_IMAGE'), initrd_image)
+        initrd_image_name = "%s-%s" % (d.getVar("INITRAMFS_IMAGE"), d.getVar("MACHINE"))
+        baseinitrd = os.path.join(d.getVar("DEPLOY_DIR_IMAGE"), initrd_image_name)
+        imgnamesuffix = d.getVar('IMAGE_NAME_SUFFIX')
+        initrd = ""
         for img in (".cpio.gz", ".cpio.lz4", ".cpio.lzo", ".cpio.lzma", ".cpio.xz", ".cpio"):
             if os.path.exists(baseinitrd + img):
                 initrd = baseinitrd + img
                 break
+            elif os.path.exists(baseinitrd + imgnamesuffix + img):
```
Comment (lumag) [inline classes/uki.bbclass:45]: If you put imgnamesuffix into play, you can skip the `baseinitrd + img` check.

File: classes/uki.bbclass [hunk 5]
```diff
@@ -33,23 +34,34 @@ python do_uki() {
...
+                initrd = baseinitrd + imgnamesuffix + img
+                break
         if not initrd:
-            bb.fatal("initrd= must be specified to create unified kernel image.")
+            bb.fatal("initrd= %s must be specified to create unified kernel image." % baseinitrd)
```
Comment (lumag) [inline classes/uki.bbclass:49]: The error message is invalid anyway. Could you please fix it instead? And the next one too.

File: classes/uki.bbclass [hunk 6]
```diff
@@ -33,23 +34,34 @@ python do_uki() {
...
         bb.fatal("linux= must be specified to create unified kernel image.")
-    ukify_cmd += " --linux=%s --uname %s" % (kernel_image, kernel_version)
+    extn = os.path.splitext(kernel_image)[1]
+    if extn == '.gz':
+        cmd = "gunzip -dc %s > %s" % (kernel_image, d.getVar('B') + "/" + "Image")
+        subprocess.check_call(cmd, shell=True)
+    else:
+        shutil.copy2(kernel_image, d.getVar('B'), follow_symlinks=True)
+    kernel_version = d.getVar('LINUX_VERSION')
+    ukify_cmd += " --linux=%s --uname %s" % (d.getVar('B') + "/" + "Image", kernel_version)
```
Comment (lumag) [inline classes/uki.bbclass:64]: This will break if `KERNEL_IMAGETYPE` is not one of `Image` / `Image.gz`.

File: classes/uki.bbclass [hunk 7]
```diff
@@ -33,23 +34,34 @@ python do_uki() {
...
-    kernel_image = d.getVar('B') + "/" + d.getVar('KERNEL_OUTPUT_DIR') + "/Image"
-    kernel_version = d.getVar('KERNEL_VERSION')
+    kernel_image = os.path.join(d.getVar('DEPLOY_DIR_IMAGE'), d.getVar('KERNEL_IMAGETYPE'))
     if not os.path.exists(kernel_image):
         bb.fatal("linux= must be specified to create unified kernel image.")
-    ukify_cmd += " --linux=%s --uname %s" % (kernel_image, kernel_version)
+    extn = os.path.splitext(kernel_image)[1]
+    if extn == '.gz':
+        cmd = "gunzip -dc %s > %s" % (kernel_image, d.getVar('B') + "/" + "Image")
+        subprocess.check_call(cmd, shell=True)
```
Comment (lumag) [inline classes/uki.bbclass:60]: It looks like this should be handled in a different way. E.g. by making sure that the user has `Image` in the `KERNEL_IMAGETYPES` instead.

File: classes/uki.bbclass [hunk 8]
```diff
@@ -59,15 +71,21 @@ python do_uki() {
     target_arch = d.getVar('EFI_ARCH')
     ukify_cmd += " --efi-arch %s" % target_arch
 
+    # OS-release
+    osrelease = os.path.join(d.getVar('RECIPE_SYSROOT'), d.getVar('libdir'), "os-release")
+    ukify_cmd += " --os-release=@%s" % osrelease
+
     # Stub
-    stub = "%s/linux%s.efi.stub" % (deploy_dir_image, target_arch)
+    stub = os.path.join(d.getVar('DEPLOY_DIR_IMAGE'), "linux%s.efi.stub" % target_arch)
```
Comment (lumag) [inline classes/uki.bbclass:79]: why?

File: classes/uki.bbclass [hunk 9]
```diff
@@ -59,15 +71,21 @@ python do_uki() {
...
     if not os.path.exists(stub):
         bb.fatal("stub must be specified to create unified kernel image %s" %stub)
     ukify_cmd += " --stub %s" % stub
 
     # Custom UKI name
-    output_dir = d.getVar('D') + "/" +  d.getVar('EFI_UKI_PATH')
-    os.makedirs(output_dir, exist_ok=True)
+    output_dir = d.getVar('B') + "/uki/" +  d.getVar('EFI_UKI_PATH')
+    if (os.path.exists(output_dir)):
```
Comment (lumag) [inline classes/uki.bbclass:86]: Please use existing bitbake mechanism for this

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 10]
```diff
@@ -0,0 +1,22 @@
+DESCRIPTION = "Qualcomm linux kernel UKI creation"
+
+LICENSE = "BSD-3-Clause-Clear"
+LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/BSD-3-Clause-Clear;md5=7a434440b651f4a472ca93716d01033a"
+
+COMPATIBLE_HOST = '(x86_64.*|arm.*|aarch64.*)-(linux.*)'
+
+inherit uki
+
+do_configure[noexec] = "1"
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:10]: Why? Default do_configure should just fall through.

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+DESCRIPTION = "Qualcomm linux kernel UKI creation"
+
+LICENSE = "BSD-3-Clause-Clear"
+LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/BSD-3-Clause-Clear;md5=7a434440b651f4a472ca93716d01033a"
+
+COMPATIBLE_HOST = '(x86_64.*|arm.*|aarch64.*)-(linux.*)'
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:6]: If the package is called `linux-qcom-uki`, it doesn't make sense to list `x86_64` as compatible. I think it makes sense to rename the recipe to a generic one and inherently send it to OE-core for inclusion.

File: conf/machine/include/qcom-common.inc [hunk 12]
```diff
@@ -40,6 +40,7 @@ INITRAMFS_MAXSIZE = "393216"
 
 # Use systemd-boot as the EFI bootloader
 EFI_PROVIDER = "systemd-boot"
+PREFERRED_RPROVIDER_virtual-systemd-bootconf = "qcom-systemd-bootconf"
```
Comment (lumag) [inline conf/machine/include/qcom-common.inc:43]: Commit messages is not finished. Also this doesn't seem relevant to the INITRAMFS_IMAGE issue.

File: recipes-core/systemd/qcom-systemd-bootconf.bb [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/BSD-3-Clause-Clear;md5=7a434440b651f4a472ca93716d01033a"
+
+do_configure[noexec] = "1"
+do_compile[noexec] = "1"
+
+do_install() {
+    # TODO: Create it similar to systemd-bootconf
+    mkdir -p ${D}${EFI_PREFIX}/loader
+    touch ${D}${EFI_PREFIX}/loader/loader.conf
+}
```
Comment (lumag) [inline recipes-core/systemd/qcom-systemd-bootconf.bb:13]: Why do you need it at all?

File: recipes-kernel/images/esp-qcom-image.bb [hunk 14]
```diff
@@ -17,3 +17,11 @@ IMAGE_FSTYPES = "vfat"
 IMAGE_FSTYPES_DEBUGFS = ""
 
 LINGUAS_INSTALL = ""
+
+IMAGE_ROOTFS_SIZE = "4096"
+IMAGE_ROOTFS_EXTRA_SPACE = "512"
```
Comment (lumag) [inline recipes-kernel/images/esp-qcom-image.bb:22]: Again, irrelevant to INITRAMFS_IMAGE issue.

**Review by lumag — CHANGES_REQUESTED** [review]
Also just to point it out, starting from commit 4 you have switched to the `qti.qualcomm.com`. Please fix your git configuration and update the patches.

File: recipes-core/systemd/systemd-boot-native_254.4.bb [hunk 15]
```diff
@@ -1,15 +0,0 @@
...
-
-COMPATIBLE_HOST = "(aarch64.*|arm.*|x86_64.*|i.86.*)-linux"
-
-do_configure[noexec] = "1"
-do_compile[noexec] = "1"
-
-do_install () {
-	install -d ${D}${bindir}
-	install -m 0755 ${S}/src/ukify/ukify.py ${D}${bindir}/ukify
-}
```
[the remark this answers is not in the document]
Comment (quic-vkraleti) [reply recipes-core/systemd/systemd-boot-native_254.4.bb:15]: Ok, I'll submit it separately.

File: recipes-kernel/images/linux-qcom-uki.bb (hunk 1 above)
Comment (quic-vkraleti) [reply recipes-kernel/images/linux-qcom-uki.bb:14]: Done.

File: classes/uki.bbclass (hunk 3 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:22]: fixed.

File: classes/uki.bbclass (hunk 4 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:45]: Done.

File: classes/uki.bbclass (hunk 6 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:64]: For now baling out if kernel image type is not Image/Image.gz. Will explore if there is a possibility to convert zImage, uImage, bzImage etc into Image.

File: classes/uki.bbclass (hunk 7 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:60]: Sure, will explore.

File: classes/uki.bbclass (hunk 8 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:79]: to get rid of deploy_dir_image local variable.

Comment (lumag) [reply classes/uki.bbclass:79]: 1) Unrelated changes make it harder to read the patch.
2) There is nothing wrong with local variables.

File: recipes-kernel/images/linux-qcom-uki.bb (hunk 11 above)
Comment (quic-vkraleti) [reply recipes-kernel/images/linux-qcom-uki.bb:6]: Sure will do. For now dropped x86_64 from compatible host list.

File: classes/uki.bbclass (hunk 6 above)
Comment (lumag) [reply classes/uki.bbclass:64]: This is not what I have asked for.

File: recipes-kernel/images/esp-qcom-image.bb (hunk 14 above)
Comment (quic-vkraleti) [reply recipes-kernel/images/esp-qcom-image.bb:22]: will move to another pull request.

File: recipes-core/systemd/qcom-systemd-bootconf.bb (hunk 13 above)
Comment (lumag) [reply recipes-core/systemd/qcom-systemd-bootconf.bb:13]: This doesn't answer the question. The "is not serving needs" is not a sensible response. I can not make any judgement using it.

Comment (quic-vkraleti) [reply recipes-core/systemd/qcom-systemd-bootconf.bb:13]: Will provide more details on what problems we faced in next pull request we are going to submit for this.

File: classes/uki.bbclass (hunk 6 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass]: Added Image to KERNEL_IMAGETYPES

File: recipes-kernel/images/linux-qcom-uki.bb (hunk 10 above)
Comment (quic-vkraleti) [reply recipes-kernel/images/linux-qcom-uki.bb]: Done

File: classes/uki.bbclass [hunk 16]
```diff
@@ -13,17 +13,20 @@ DEPENDS:append = " \
...
-do_install[depends] += " \
+do_configure[depends] += " \
+    os-release:do_populate_sysroot \
     systemd-boot:do_deploy \
+    virtual/kernel:do_deploy \
     "
-do_install[depends] += "${@ '${INITRAMFS_IMAGE}:do_image_complete' if d.getVar('INITRAMFS_IMAGE') else ''}"
-do_uki[dirs] = "${B}"
-do_install[postfuncs] += "do_uki"
+do_configure[depends] += "${@ '${INITRAMFS_IMAGE}:do_image_complete' if d.getVar('INITRAMFS_IMAGE') else ''}"
```
Comment (lumag) [inline classes/uki.bbclass:23]: this is just `DEPENDS = "os-release systemd-boot virtual/kernel"`

File: classes/uki.bbclass [hunk 17]
```diff
@@ -76,15 +97,6 @@ python do_uki() {
     env['LD_LIBRARY_PATH'] = d.expand("${RECIPE_SYSROOT_NATIVE}/usr/lib/systemd:${LD_LIBRARY_PATH}")
 
     # Run the ukify command
+    bb.debug(1, "ukify_cmd: %s" % ukify_cmd)
     subprocess.check_call(ukify_cmd, env=env, shell=True)
```
Comment (lumag) [inline classes/uki.bbclass:101]: This funcion now looks completely like a shell script rewritten into python. Can we please rewrite it back to shell?

File: conf/machine/include/qcom-common.inc [hunk 18]
```diff
@@ -52,3 +52,6 @@ EFI_LINUX_IMG ?= "linux-${MACHINE}.efi"
 
 # Place dtb at EFIDTDIR to seamlessly package
 KERNEL_DTBDEST = "${EFI_DTB_DIR}"
+
+# UKI generation need uncompressed Kernel image
+KERNEL_IMAGETYPES:append = " Image"
```
Comment (lumag) [inline conf/machine/include/qcom-common.inc:57]: This is irrelevant to the linux-qcom-uki recipe. Move this to a separate commit.

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 19]
```diff
@@ -0,0 +1,16 @@
+DESCRIPTION = "Qualcomm linux kernel UKI creation"
+
+LICENSE = "BSD-3-Clause-Clear"
+LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/BSD-3-Clause-Clear;md5=7a434440b651f4a472ca93716d01033a"
+
+COMPATIBLE_HOST = '(arm.*|aarch64.*)-(linux.*)'
+
+inherit uki
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:8]: I think it is easier to drop uki.bbclass now, as we have it mostly rewritten. Just merge its contents here.

**Review by lumag — CHANGES_REQUESTED** [review]
We should skip building the linux-qcom-uki if `KERNEL_IMAGETYPES` doesn't contain `Image`. That's what I had in mind in my comment. Please use `SKIP_RECIPE` or `bb.parse.SkipRecipe`

File: classes/uki.bbclass (hunk 17 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:101]: I can but what's wrong if it is written in Python?

File: classes/uki.bbclass (hunk 16 above)
Comment (quic-vkraleti) [reply classes/uki.bbclass:23]: Not really. DEPENDS make current recipe's do_configure wait for do_recipe_sysroot of dependent recipes to complete. But here do_configure need to wait for do_deploy as systemd-boot & kernel recipes are not providing stub & Image as a package. So explicit task depends is still needed.

File: classes/uki.bbclass (hunk 17 above)
Comment (lumag) [reply classes/uki.bbclass:101]: What is wrong with having Python wrapper around shell? Probably the same as having the ineffective code in any other software piece.

File: classes/uki.bbclass (hunk 16 above)
Comment (lumag) [reply classes/uki.bbclass:23]: Ack.

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    # DTB
+    if [ -n "${UKI_DTB}" ]; then
+        dtb_path="${DEPLOY_DIR_IMAGE}/${UKI_DTB}"
+        [ -f $dtb_path ] && echo "Creating unified kernel image with $dtb_path" || die "$dtb_path is not a valid dtb to create UKI."
+        ukify_cmd="$ukify_cmd --devicetree $dtb_path"
+    fi
+
+    # Output
+    mkdir -p "${D}${EFI_UKI_PATH}"
+    output="${D}${EFI_UKI_PATH}/${EFI_LINUX_IMG}"
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:79]: Compile should buid the image, not install it.
${D} should only be touched in do_install.

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    stub="${DEPLOY_DIR_IMAGE}/linux${EFI_ARCH}.efi.stub"
+    [ -f $stub ] && echo "Creating unified kernel image with $stub" || die "$stub is not a valid stub to create UKI."
+    ukify_cmd="$ukify_cmd --stub $stub"
+
+    # DTB
+    if [ -n "${UKI_DTB}" ]; then
+        dtb_path="${DEPLOY_DIR_IMAGE}/${UKI_DTB}"
+        [ -f $dtb_path ] && echo "Creating unified kernel image with $dtb_path" || die "$dtb_path is not a valid dtb to create UKI."
+        ukify_cmd="$ukify_cmd --devicetree $dtb_path"
+    fi
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:75]: This is called `KERNEL_DEVICETREE`, isn't it?

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+do_compile() {
+    # Construct the ukify command
+    ukify_cmd=""
+
+    # Ramdisk
+    if [ -n "${INITRAMFS_IMAGE}" ]; then
+        initrd=""
+        for img in cpio cpio.gz cpio.lz4 cpio.lzo cpio.lzma cpio.xz cpio.zst; do
+            if [ -e "${DEPLOY_DIR_IMAGE}/${INITRAMFS_IMAGE}-${MACHINE}${IMAGE_NAME_SUFFIX}.$img" ]; then
+                initrd="${DEPLOY_DIR_IMAGE}/${INITRAMFS_IMAGE}-${MACHINE}${IMAGE_NAME_SUFFIX}.$img"
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:36]: No need for `-${MACHINE}` here.

File: recipes-kernel/images/linux-qcom-uki.bb [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+                break
+            fi
+        done
+        [ -f $initrd ] && echo "Creating unified kernel image with $initrd" || die "$initrd is not a valid initrd to create UKI."
```
Comment (lumag) [inline recipes-kernel/images/linux-qcom-uki.bb:40]: `bbfatal`, please.

**Review by lumag — CHANGES_REQUESTED** [review]

File: recipes-kernel/images/linux-qcom-uki.bb (hunk 21 above)
Comment (quic-vkraleti) [reply recipes-kernel/images/linux-qcom-uki.bb:75]: KERNEL_DEVICETREE can have multiple DTB files. But within UKI we want to handle only a subset of those.  More over want to keep this section optional as per our previous discussions and move all dtb files into ESP directly. Hence decided to use a separate variable here.

Comment (lumag) [reply recipes-kernel/images/linux-qcom-uki.bb:75]: Current code doesn't have `UKI_DTB`, so it should not be a part of this PR. Please submit it separately.

File: conf/machine/include/qcom-common.inc [hunk 24]
```diff
@@ -52,3 +52,6 @@ EFI_LINUX_IMG ?= "linux-${MACHINE}.efi"
 
 # Place dtb at EFIDTDIR to seamlessly package
 KERNEL_DTBDEST = "${EFI_DTB_DIR}"
+
+# UKI generation need uncompressed Kernel image
```
[the remark this answers is not in the document]
Comment (quic-vkraleti) [reply conf/machine/include/qcom-common.inc]: ack.

File: recipes-kernel/images/linux-qcom-uki.bb (hunk 22 above)
Comment (lumag) [reply recipes-kernel/images/linux-qcom-uki.bb:36]: Ah, I remember what is the issue. It should be `[ -n "${INITRAMFS_IMAGE}"]`, but after that it should be `${INITRAMFS_IMAGE_NAME}` all over the place.

Comment (lumag) [reply recipes-kernel/images/linux-qcom-uki.bb:36]: `INITRAMFS_IMAGE_NAME` is also defined in `image-artifact-names.bbclass`:
```
layers/openembedded-core/meta/classes-recipe/image-artifact-names.bbclass:INITRAMFS_IMAGE_NAME ?= "${@['${INITRAMFS_IMAGE}${IMAGE_MACHINE_SUFFIX}', ''][d.getVar('INITRAMFS_IMAGE') == '']}"
```

Comment (quic-vkraleti) [reply recipes-kernel/images/linux-qcom-uki.bb:36]: Aah. I am checking on Kirkstone. This got added in Master. I am clear now and updated the change. In similar lines should we set ${IMAGE_NAME_SUFFIX} to empty for initramfs-test-full-image? Expectation is all initramfs images should set to empty.

Comment (lumag) [reply recipes-kernel/images/linux-qcom-uki.bb]: Yes, please.

**Review by lumag — APPROVED** [review]

**Pull request merged**